pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV4.java

```diff
@@ -19,36 +19,306 @@
 
 package org.apache.pinot.core.common.datatable;
 
+import java.io.DataOutputStream;
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.util.HashMap;
 import java.util.Map;
 import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.datablock.DataBlockUtils;
 import org.apache.pinot.core.common.datablock.RowDataBlock;
 import org.apache.pinot.spi.annotations.InterfaceStability;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
 
 
 /**
- * Datatable V4 Implementation is a wrapper around the Row-based data block.
+ * Datatable V4 implementation is a wrapper around the Row-based data block.
+ * The layout of serialized V4 datatable looks like:
+ * +-----------------------------------------------+
+ * | 17 integers of header:                        |
+ * | VERSION                                       |
+ * | NUM_ROWS                                      |
+ * | NUM_COLUMNS                                   |
+ * | EXCEPTIONS SECTION START OFFSET               |
+ * | EXCEPTIONS SECTION LENGTH                     |
+ * | DICTIONARY_MAP SECTION START OFFSET           |
+ * | DICTIONARY_MAP SECTION LENGTH                 |
+ * | DATA_SCHEMA SECTION START OFFSET              |
+ * | DATA_SCHEMA SECTION LENGTH                    |
+ * | FIXED_SIZE_DATA SECTION START OFFSET          |
+ * | FIXED_SIZE_DATA SECTION LENGTH                |
+ * | VARIABLE_SIZE_DATA SECTION START OFFSET       |
+ * | VARIABLE_SIZE_DATA SECTION LENGTH             |
+ * | FIXED_SIZE_NULL_VECTOR SECTION START OFFSET   |
```

Comment:
We can put the bitmaps into existing `fixedSizeData` and `variableSizeData`. If no bitmap is provided, we don't write anything. If some bitmaps are provided, we store 2 ints per column in the fixed size data (variable size data buffer offset & length). On the reader side, we can derive whether null bitmap is available by checking the size of the fixed size data. The overhead would be minimal if null handling is not enabled, where the data table will be exactly the same as before.